## Final Project Submission

Please fill out:
* Student name: Laura Lewis
* Student pace: full time
* Scheduled project review date/time: 8 February 2019, 4pm GMT/11am EST
* Instructor name: Joe San Pietro
* Blog post URL:


***
**You'll clean, explore, and model this dataset with a multivariate linear regression to predict the sale price of houses as accurately as possible.**

**Based on the results of your models, your presentation should discuss at least two concrete features that highly influence housing prices.**

**Go through the Jupyter Notebook, answering questions about how you made certain decisions. Be ready to explain things like:**
* "how did you pick the question(s) that you did?"
* "why are these questions important from a business perspective?"
* "how did you decide on the data cleaning options you performed?"
* "why did you choose a given method or library?"
* "why did you select those visualizations and what did you learn from each of them?"
* "why did you pick those features as predictors?"
* "how would you interpret the results?"
* "how confident are you in the predictive quality of the results?"
* "what are some of the things that could cause the results to be wrong?"

### Technical Report Must-Haves

For this project, your Jupyter Notebook should meet the following specifications:

#### Organization/Code Cleanliness

* The notebook should be well organized, easy to follow,  and code should be commented where appropriate.  
    * Level Up: The notebook contains well-formatted, professional looking markdown cells explaining any substantial code.  All functions have docstrings that act as professional-quality documentation
* The notebook is written for a technical audiences with a way to both understand your approach and reproduce your results. The target audience for this deliverable is other data scientists looking to validate your findings. 

#### Visualizations & EDA

* Your project contains at least 4 _meaningful_ data visualizations, with corresponding interpretations. All visualizations are well labeled with axes labels, a title, and a legend (when appropriate)  
* You pose at least 3 meaningful questions and aswer them through EDA.  These questions should be well labled and easy to identify inside the notebook. 
    * **Level Up**: Each question is clearly answered with a visualization that makes the answer easy to understand.   
* Your notebook should contain 1 - 2 paragraphs briefly explaining your approach to this project **through the OSEMN framework**. 
    
#### Model Quality/Approach

* Your model should not include any predictors with p-values greater than .05.  
* Your notebook shows an iterative approach to modeling, and details the parameters and results of the model at each iteration.  
    * **Level Up**: Whenever necessary, you briefly explain the changes made from one iteration to the next, and why you made these choices.  
* You provide at least 1 paragraph explaining your final model.   
* You pick at least 3 coefficients from your final model and explain their impact on the price of a house in this dataset.   
***

**Project plan (OSEMN framework):**

Obtain:

- Import data using pandas read.csv
- Research and understand variables

Scrub:

- Null values
- Outliers
- Missing values - drop, keep or impute
- Removing rows or columns
- Converting formats

Explore:
- Transformation - standardisation and normalisation

Model:

Interpret:

# Project outline

Plan - OSEMN

Overall aim - predict house prices in...

Questions - 1, 2, 3

***
# Obtaining the data

### _Importing the data_

In this section, the data (and libraries) will be imported and inspected to see if any initial adjustments are necessary e.g. changing the index. Any necessary research into the variables will also be conducted.

In [1]:
# Importing libraries to be used in this project
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import Normalizer
from sklearn.preprocessing import StandardScaler
import statsmodels.stats.api as sms
import statsmodels.api as sm
from statsmodels.formula.api import ols
import scipy.stats as stats
plt.style.use('ggplot')
% matplotlib inline

In [2]:
kc = pd.read_csv('kc_house_data.csv') # Import csv
pd.set_option('display.max_columns', None) # Displays all columns. Can be reset with: pd.reset_option('display.max_columns')
kc.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0         NaN   0.0          3      7        1180   
1      7242     2.0         0.0   0.0          3      7        2170   
2     10000     1.0         0.0   0.0          3      6         770   
3      5000     1.0         0.0   0.0          5      7        1050   
4      8080     1.0         0.0   0.0          3      8        1680   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   
3         910.0      1965           0.0    98136  47.5208 -122.393   
4           0.0      1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

In [3]:
kc.tail()

id        date     price  bedrooms  bathrooms  sqft_living  \
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
21592      1131     3.0         0.0   0.0          3      8        1530   
21593      5813     2.0         0.0   0.0          3      8        2310   
21594      1350     2.0         0.0   0.0          3      7        1020   
21595      2388     2.0         NaN   0.0          3      8        1600   
21596      1076     2.0         0.0   0.0          3      7        1020   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
21592           0.0      2009           0.0    98103  47.6993 -122.346   
21593           0.0      2014           0.0    98146  47.5107 -122.362   
21594           0.0      2009           0.0    98144  47.5944 -122.299   
21595           0.0      2004           0.0    98027  47.5345 -122.069   
21596           0.0      2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357

An id field is already included. A quick check can be run using value_counts to see whether these are unique values and could therefore be used as the index:

In [4]:
kc.id.value_counts().head()

795000620     3
1825069031    2
2019200220    2
7129304540    2
1781500435    2
Name: id, dtype: int64

In [ ]:
kc.loc[kc['id'] == 795000620] # Checking one example of repeated id values

id        date     price  bedrooms  bathrooms  sqft_living  \
17588  795000620   9/24/2014  115000.0         3        1.0         1080   
17589  795000620  12/15/2014  124000.0         3        1.0         1080   
17590  795000620   3/11/2015  157000.0         3        1.0         1080   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
17588      6250     1.0         0.0   0.0          2      5        1080   
17589      6250     1.0         0.0   0.0          2      5        1080   
17590      6250     1.0         NaN   0.0          2      5        1080   

      sqft_basement  yr_built  yr_renovated  zipcode      lat    long  \
17588           0.0      1950           0.0    98168  47.5045 -122.33   
17589           0.0      1950           0.0    98168  47.5045 -122.33   
17590           0.0      1950           NaN    98168  47.5045 -122.33   

       sqft_living15  sqft_lot15  
17588           1070        6250  
17589           1070        6250  
17590           1070        6250

Some id numbers are repeated because the same house was sold on several different dates, so this is not suitable as a primary key/index. It can therefore be dropped, as it does not give us any useful information for the purposes of this analysis.

In [ ]:
kc.drop(['id'], axis=1, inplace=True)

### _Understanding the variables_

#### Column names and descriptions for Kings County Data Set, from `column_names.md`:
* **id** - unique identified for a house
* **dateDate** - house was sold
* **pricePrice** -  is prediction target
* **bedroomsNumber** -  of Bedrooms/House
* **bathroomsNumber** -  of bathrooms/bedrooms
* **sqft_livingsquare** -  footage of the home
* **sqft_lotsquare** -  footage of the lot
* **floorsTotal** -  floors (levels) in house
* **waterfront** - House which has a view to a waterfront
* **view** - Has been viewed
* **condition** - How good the condition is ( Overall )
* **grade** - overall grade given to the housing unit, based on King County grading system
* **sqft_above** - square footage of house apart from basement
* **sqft_basement** - square footage of the basement
* **yr_built** - Built Year
* **yr_renovated** - Year when house was renovated
* **zipcode** - zip
* **lat** - Latitude coordinate
* **long** - Longitude coordinate
* **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
* **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors

It is not immediately obvious what condition and grade refer to, and so these require further investigation. The King County website (https://info.kingcounty.gov/assessor/esales/Glossary.aspx?type=r) gives the following definitions. To summarise, building condition is rated on a 1-5 scale where 5 is highest, and building grade is rated on a 1-13 scale where 13 is highest.

**BUILDING CONDITION**

Relative to age and grade. Coded 1-5.

1 = Poor- Worn out. Repair and overhaul needed on painted surfaces, roofing, plumbing, heating and numerous functional inadequacies. Excessive deferred maintenance and abuse, limited value-in-use, approaching abandonment or major reconstruction; reuse or change in occupancy is imminent. Effective age is near the end of the scale regardless of the actual chronological age.

2 = Fair- Badly worn. Much repair needed. Many items need refinishing or overhauling, deferred maintenance obvious, inadequate building utility and systems all shortening the life expectancy and increasing the effective age.

3 = Average- Some evidence of deferred maintenance and normal obsolescence with age in that a few minor repairs are needed, along with some refinishing. All major components still functional and contributing toward an extended life expectancy. Effective age and utility is standard for like properties of its class and usage.

4 = Good- No obvious maintenance required but neither is everything new. Appearance and utility are above the standard and the overall effective age will be lower than the typical property.

5= Very Good- All items well maintained, many having been overhauled and repaired as they have shown signs of wear, increasing the life expectancy and lowering the effective age with little deterioration or obsolescence evident with a high degree of utility. 

**BUILDING GRADE**

Represents the construction quality of improvements. Grades run from grade 1 to 13. Generally defined as:

1-3 Falls short of minimum building standards. Normally cabin or inferior structure.

4 Generally older, low quality construction. Does not meet code.

5 Low construction costs and workmanship. Small, simple design.

6 Lowest grade currently meeting building code. Low quality materials and simple designs.

7 Average grade of construction and design. Commonly seen in plats and older sub-divisions.

8 Just above average in construction and design. Usually better materials in both the exterior and interior finish work.

9 Better architectural design with extra interior and exterior design and quality.

10 Homes of this quality generally have high quality features. Finish work is better and more design quality is seen in the floor plans. Generally have a larger square footage.

11 Custom design and higher quality finish work with added amenities of solid woods, bathroom fixtures and more luxurious options.

12 Custom design and excellent builders. All materials are of the highest quality and all conveniences are present.

13 Generally custom designed and built. Mansion level. Large amount of highest quality cabinet work, wood trim, marble, entry ways etc. 

***
# Scrubbing the data

In this section, the data will be pre-processed. This will include looking for null values, missing values, incorrect data types (e.g. numbers stored as strings, or categorical data stored as integers) and multicollinearity, and taking actions including dropping rows or columns, imputing values and casting data types.

### _Checking data types and null values_

#### Initial inspection

In [ ]:
kc.info() # Inspect meta-data for the dataset, to see data types and null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 20 columns):
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(10), object(2)
memory usage: 3.3+ MB


In [ ]:
kc.describe() # Descriptive statistics for numerical columns

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  2.159700e+04   
mean   5.402966e+05      3.373200      2.115826   2080.321850  1.509941e+04   
std    3.673681e+05      0.926299      0.768984    918.106125  4.141264e+04   
min    7.800000e+04      1.000000      0.500000    370.000000  5.200000e+02   
25%    3.220000e+05      3.000000      1.750000   1430.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068500e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21597.000000  19221.000000  21534.000000  21597.000000  21597.000000   
mean       1.494096      0.007596      0.233863      3.409825      7.657915   
std        0.539683      0.086825      0.765686      0.650546      1.173200   
min        1.000000      0.000000      0.000000      1.000000      3.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above      yr_built  yr_renovated       zipcode           lat  \
count  21597.000000  21597.000000  17755.000000  21597.000000  21597.000000   
mean    1788.596842   1970.999676     83.636778  98077.951845     47.560093   
std      827.759761     29.375234    399.946414     53.513072      0.138552   
min      370.000000   1900.000000      0.000000  98001.000000     47.155900   
25%     1190.000000   1951.000000      0.000000  98033.000000     47.471100   
50%     1560.000000   1975.000000      0.000000  98065.000000     47.571800   
75%     2210.000000   1997.000000      0.000000  98118.000000     47.678000   
max     9410.000000   2015.000000   2015.000000  98199.000000     47.777600   

               long  sqft_living15     sqft_lot15  
count  21597.000000   21597.000000   21597.000000  
mean    -122.213982    1986.620318   12758.283512  
std        0.140724     685.230472   27274.441950  
min     -122.519000     399.000000     651.000000  
25%     -122.328000    1490.000000    5100.000000  
50%     -122.231000    1840.000000    7620.000000  
75%     -122.125000    2360.000000   10083.000000  
max     -121.315000    6210.000000  871200.000000

In [ ]:
kc.isnull().sum() # Checking the number of null values in each column

date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

An initial inspection reveals that there are 21,597 rows in the dataset. The dataset is not so large that it is worth sub-setting to speed up processing time.

There are 8 variables classed as floats, 11 as integers and 2 as string.

There are three ways that null or unknown values might be stored - as missing (null) values, or with a placeholder for unknown values (e.g. 'NA', 'Nan') entered as a string, or in numerical data as an unlikely numerical category. The view and yr_renovated variables contain the former.

#### Changes to data types and dealing with null values:

_Date_ - cast to datetime. Can possibly later be reformatted to month and year, to allow for the assessment of whether there are any significant changes in price over time that need to be accounted for - but the day can probably be dropped as this level of detail is unnecessary given that this project will not include time series/seasonality analysis.

In [ ]:
kc.date = pd.to_datetime(kc.date) # Using the pandas 'to_datetime' function to cast the string as a date
kc.head()

_Bathrooms_ - presumably a float because .5 of a bathroom refers to a toilet and sink but no bath/shower, and 0.25 of a bathroom refers to just a toilet. Checking the unique values reveals that this is correct, and only legitimate decimals (.25, .5 and .75 for various bathroom fitting combinations) are used.

In [ ]:
kc.bathrooms.unique() # Checking which unique values for bathroom appear

_Floors_ - currently stored as a float, but should it be an integer? However, after checking the unique values it seem that the only decimals used are .5 mezzanines are used, so this is ok to leave as a float.

In [ ]:
kc.floors.unique() # Checking which unique values for floors appear

_Waterfront_ - currently stored as a float. Checking the values reveals that this is a boolean variable with 1s and 0s (and some null values), and so it should be cast to a string rather than storing it as a float. Null values were replaced with 0 (the median and mode).

In [ ]:
kc.waterfront.value_counts() # Checking category counts

In [ ]:
kc.waterfront.isna().sum()

In [ ]:
kc.waterfront.fillna(0, inplace=True) # Replace all null values in waterfront with 0

In [ ]:
kc.waterfront.value_counts()

In [ ]:
kc.waterfront = kc.waterfront.astype('str') # Casting the waterfront column to a string

_View_ - currently stored as a float, but defined as 'has been viewed' which implies a boolean variable for whether or not a property has been viewed. However, checking the unique values reveals that this is not the case - values are 0, 1, 2, 3, 4 and nan. It could be the number of times a property has been viewed, but the low max and cardinality suggest not. The most likely explanation is that it is another integer grading system similar to condition and grade. Research online for other analyses of this dataset provides both definitions, but the latter (a grading of the view quality from 0-4, with 4 being the highest) is the most commonly accepted and will be used. Counts of rows in each view category reveals that the vast majority (90%) of entries have a view of 0. Null values will be replaced with 0, and the datatype will be cast to integer.

In [ ]:
kc.view.unique() # Checking which unique values for view appear

In [ ]:
print(kc.view.value_counts()) # Counts of rows for each possible value for view
print((kc.view.value_counts().head(1)/kc.view.value_counts().sum())*100) # Percentage of rows in the most common view category

In [ ]:
kc.view.fillna(0, inplace=True)

In [ ]:
kc.view.value_counts()

In [ ]:
kc['view'] = kc['view'].astype('int64')

*Sqft_basement* - should be an integer (or possibly float) as with other sqft measurements. However, attempting to cast it to an integer produces an error, because some values contain text (making this a string variable). The row counts for each unique value reveal that 454 rows have a sqft_basment value of '?', hence this error. This represents 2.1% of data, so it is not worth removing the column. The options for dealing with this are: drop the rows containing '?', replace them with the column average (in this case a value of 0 would be most appropriate, as this is over half of the values and so would be both the mode and the median), or use coarse classification to bin the data and use ? as a category. Dropping data loses data and binning data reduces granularity of data. In this case, given that the majority of entries contain the same value (0), it is relatively safe to replace missing values with 0.

In [ ]:
kc.sqft_basement.value_counts().sort_values(ascending=False) # Counts of rows for each possible value of sqft_basement

In [ ]:
kc.loc[kc['sqft_basement'] == '?', 'sqft_basement'] = 0 # Replace all values of '?' with 0
kc.sqft_basement = kc.sqft_basement.astype('float').astype('int64') # Cast to integer (via float to deal with '0.0')
kc.sqft_basement.dtype # Confirm type is correctly cast

*Yr_renovated* - should be an integer rather than a float, as with yr_built. However, there are 3,842 null values. This would be too many rows to remove from the dataset (17.8%). It is not possible to replace them with the mean year, because this could result in renovation years being prior to build years, which is impossible. Binning would reduce the granularity of data. One option would be to replace the null values with the year that represents the average amount of time between building and renovating a property. However, the easiest solution in this option is to replace the null values with the mode value 0.

In [ ]:
kc.yr_renovated.unique()

In [ ]:
kc.yr_renovated.isna().sum() # Number of null values

In [ ]:
kc.yr_renovated.value_counts().sort_values(ascending=False)

In [ ]:
kc.yr_renovated.fillna(0, inplace=True) # Replace all null values in yr_renovated with 0
kc.yr_renovated = kc.yr_renovated.astype('int64') # Casting the yr_renovated column to an integer
kc.yr_renovated.dtype # Confirm type is correctly cast

_Zipcode_ - currently stored as an integer, but should be a string as these is a categorical rather than a continuous variable. A check of unique values reveals no '?'s or other missing value placeholders.

In [ ]:
kc.zipcode = kc.zipcode.astype('str')

In [ ]:
kc.zipcode.unique()

Some of the sqft measurements might be related to each other, which could cause problems in modelling because it would result in two variables producing the same effect. The main possibility is that sqft_living is the sum of sqft_above and sqft_basement. However, a comparison reveals that this is not always the case, and that there are some discrepancies in numbers. Therefore sqft_living will not be dropped as these discrepancies may be useful information (although it is also possible that they are errors).

In [ ]:
kc['sqft_living2'] = kc.sqft_above + kc.sqft_basement

In [ ]:
kc['sqft_diff'] = kc.sqft_living - kc.sqft_living2

In [ ]:
kc.sqft_diff.value_counts().head()

In [ ]:
kc.sqft_living.equals(kc.sqft_living2)

In [ ]:
kc.drop('sqft_living2', axis=1, inplace=True)
kc.drop('sqft_diff', axis=1, inplace=True)

A final check of data types:

In [ ]:
kc.info()

A final check for null values:

In [ ]:
kc.isna().sum().sum()

### _Checking for multicollinearity_

In [ ]:
# Set the style of the visualization
sns.set(style="white")

# Create a covariance matrix
corr = kc.corr()

# Generate a mask the size of our covariance matrix
mask = np.zeros_like(corr, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True

# Set up the matplotlib figure
f, ax = plt.subplots(figsize = (11,9))

# Generate a custom diverging colormap
cmap = sns.diverging_palette(220, 10, as_cmap=True)

# Draw the heatmap with the mask and correct aspect ratio
sns.heatmap(corr, mask=mask, cmap=cmap, vmax=.3, center=0, square=True, linewidths=.5, cbar_kws={"shrink": .5});

There are quite a few columns that are mildly correlated with each other, including several that appear to be share about 30% of their variation with the target variable (price). However, no variables are so strongly correlated that they need to be removed.

***
# Exploring the data

In this section, exploratory data analysis will be conducted, as will any additional pre-processing/cleaning tasks that become apparently. This will include inspecting distributions, checking for outliers in continuous numerical data, transformation (standardisation/normalisation), one-hot encoding, and checking the dataset meets the assumptions required for linear regression.

### _Inspecting distributions and conducting transformations_

The histograms below show the distributions for each variable.

In [ ]:
kc.hist(figsize=(15,15));

#### Initial visual inspection of distributions:

- Bathrooms and grade are fairly normally distributed
- Bedrooms has some high outliers which require additional inspection and possibly log transformation
- Condition, floors and view have only a small number of groupings and are not normally distributed
- Latitude is roughly normally distributed but negatively skewed
- Longitude, sqft_above, sqft_basement, sqft_living and sqft_living15 are roughly normally distributed but positively skewed, and may require log transformation
- Price is heavily positively skewed and will require log transformation
- Sqft_lot, sqft_lot15 and yr_renovated appear to have some high outliers which require additional inspection and possibly log transformation
- Yr_built is not very normally distributed, and is negatively skewed

In [ ]:
# Examining the distributions via boxplots for the heavily positively skewed distributions above, to see if there are any extreme outliers
for col in ['bedrooms', 'price', 'sqft_above', 'sqft_basement', 'sqft_living', 'sqft_lot', 'sqft_lot15']:
    plt.boxplot(kc[col])
    plt.title(col)
    plt.show()

The boxplots above show that most of the variables plotted have very long tails, i.e. there are a very small number of very large properties. The main candidate for an actual outlier is one property with 33 bedrooms. As this is three times the number of bedrooms of the next largest property (and enough bedrooms to be a hotel), this may be data entry error, e.g. someone accidentally entering 3 instead of 33. Looking up the rest of the data for this house (below), this does appear to be a data entry error - a 33 bedroom property could not have 1.75 bathrooms or fit into 1620 square feet on one floor. Therefore it is reasonable to change this to a bedroom number of 3.

In [ ]:
kc.bedrooms.value_counts() # Checking the value for the property with a very large number of bedrooms

In [ ]:
kc.loc[kc.bedrooms == 33]

In [ ]:
kc.loc[kc.bedrooms == 33, 'bedrooms'] = 3

In [ ]:
kc.loc[kc.index == 15856] # Confirm change

#### Log-transformation:

Potential candidates for log-transformation are: bathrooms, bedrooms, grade, long, price, sqft_above, sqft_basement, sqft_living, sqft_living15, sqft_lot, sqft_lot15, yr_renovated. However, sqft_basement and yr_renovated contain values of 0, and longitude contains negative numbers, so they cannot be log transformed. A new dataframe with columns for log-transformed versions of these variables will be created.

In [ ]:
kc.describe()

In [ ]:
kc.head()

In [ ]:
kc_log = kc.drop(['date', 'zipcode'], axis=1) # Dropping date and zipcode as these will not be transformed
for col in ['bathrooms', 'bedrooms', 'grade', 'price', 'sqft_above', 'sqft_living', 'sqft_living15', 'sqft_lot', 'sqft_lot15']:
    kc_log[col] = np.log(kc_log[col])

#### Feature scaling:

The variables are still on different order of magnitude, so some form of feature scaling is required. MinMax scaling using sklearn will be used, in order to coerce each variable into the range 0-1.

In [ ]:
kc_log.describe()

In [ ]:
kc_log.describe().columns # Get column names for transformation

In [ ]:
# Min-max scale variables
transformer = MinMaxScaler()
kc_log[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built',
       'yr_renovated', 'lat', 'long', 'sqft_living15', 'sqft_lot15']] = transformer.fit_transform(kc_log[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built',
       'yr_renovated', 'lat', 'long', 'sqft_living15', 'sqft_lot15']])

In [ ]:
kc_log.describe() # Confirm min-max transformation

Displots are plotted below using seaborn, to assess the distributions and normality of the newly-transformed data. The target variable price is now normally distributed. The predictor variables sqft_living, sqft_lot, grade, sqft_above, sqft_living15 and sqft_lot15 also now look normally distributed, and are strong candidates for good explanatory variables in the modeling stage.

In [ ]:
# Plotting sns.distplots for log-transformed variables 
for col in kc_log[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built',
       'yr_renovated', 'lat', 'long', 'sqft_living15', 'sqft_lot15']]:
    plt.figure()
    sns.distplot(kc_log[col])
    plt.show()

Yr_renovated will be dropped as there are so many 0 values in this variable that it will not be possible to transform it into being normally distributed.

In [ ]:
kc_log.drop(['yr_renovated'], axis=1, inplace=True)

### _One-hot encoding_

At this point, the categorial variable waterfront can be one-hot encoded, to produce columns for 1 (yes) and 0 (no).

In [ ]:
waterfront_dummies = pd.get_dummies(kc_log['waterfront'], prefix='waterfront') # Get dummies
kc_log.drop(['waterfront'], axis=1, inplace=True) # Drop original waterfront column
kc_log = pd.concat([kc_log, waterfront_dummies], axis=1) # Add dummy variables to original dataset
kc_log.rename(columns={'waterfront_0.0':'waterfront_no', 'waterfront_1.0':'waterfront_yes'})
kc_log.head()

### _Checking the assumptions for linear regression_

In order to conduct regression analysis, certain assumptions must be met. The normality assumption has already been partly tested above using histograms and distplots, and this will be confirmed with Q-Q plots and the Jarque-Bera test. The linearity assumption between the predictor variables and the target variable will also be tested, as will heteroscedasticity (including with the Goldfeld-Quandt test).

#### Normality assumption:

This will be tested with Q-Q plots and the Jarque-Bera (JB) test.

The Q-Q plots below and the results of the JB test reveal that none of the variables are normally distributed. However, for now we shall proceed with these variables to see whether a regression model can be made with any degree of predictive power, and this can be iterated on later. The top 5 most normally-distributed variables are sqft_living, sqft_above, grade, bathrooms and view.

In [ ]:
results = [['independent_variable', 'r_squared', 'intercept', 'slope', 'p-value', 'normality(JB)']]

for i, col in enumerate(['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'lat', 'long', 'sqft_living15', 'sqft_lot15', 'waterfront_1.0']):

    #Creating the OLS regression model 
    f = 'price~' + col
    model = ols(formula=f, data=kc_log).fit()
    print('Regression diagnostics for ' + f)
    
    # Plotting the residuals
    fig = plt.figure(figsize=(12,8))
    sm.graphics.plot_regress_exog(model, col, fig=fig)
    plt.show()

    # Q-Q plot
    sm.graphics.qqplot(model.resid, dist=stats.norm, line='45', fit=True)
    plt.show()
    
    # Add results from each iteration as a new list in results, so that it can be turned into a pandas dataframe with
    # the first row as the column names
    results.append([col, model.rsquared, model.params[0], model.params[1], model.pvalues[1], sms.jarque_bera(model.resid)[0]])

In [ ]:
results = pd.DataFrame(results)
results.columns = results.iloc[0]
results = results.drop(results.index[0])
results.sort_values(by=['normality(JB)'])

#### Linearity assumption and heteroscedasticity

This will be tested with sns jointplots.

The scatterplots below confirm that there is a lot of variation in the data, and that no one variable is a solid predictor of house price. Some variables exhibit a degree of heteroscedasticity, e.g. longitude, but most appear sufficiently homoscedastic to use in linear regression.

In [ ]:
for col in ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built',
       'lat', 'long', 'sqft_living15', 'sqft_lot15']:
    sns.jointplot(x= kc_log[col], y= kc_log['price'], data=kc_log, kind='reg', label=col)
    plt.legend()
    plt.show();

***
# Questions

Exploratory data analysis of the dataset brings up some interesting questions that can be explored further. Three questions will be addressed in the following section:
- How much does house price vary by location? Specifically, what is the proportional difference in price per square foot between the cheapest and the most expensive area?
- How much does having a view increase house price by?/Which has a bigger impact on house price - being on the waterfront, or having a good view?
- 

Possible questions:
- How have house prices changed over time?

- How much does being on the waterfront increase house price by? (compare prices for waterfront 1 and waterfront 0, and also compare other values/try and hold them constant)
- How much does having a view increase your house price by? (compare prices for 0 view and numbered views)

- How much does the quality of a house (condition and grade) affect the price for a house of a given size (number of bedrooms)?
- Something to do with how much bigger or smaller a house is compared to neighbours (take square footage away from average square footage of neighbours)

- What is the best single predictor of house price? (answer with single-variable linear regression)

### Question 1: how much does house price vary by location?

It is generally known that location is a very important factor in house price. The over-arching question of how much house price varies by location will be answered specifically in the form of: what is the proportional difference in price per square foot between the cheapest and the most expensive area?

A 2-dimensional histogram with longitude on the x axis and latitude on the y axis can be used to produce a map of where exactly in King County the properties area. This figure shows that a lot of properties are found in and around Seattle, but that the dataset also includes properties further away from the city. Lot size is likely to vary in a non-random way, with smaller lots in the city and larger lots in the country. Therefore, a better metric to compare like-for-like may be to compare price per square foot (ppsf).

In [ ]:
# Plotting a 2D histogram
plt.figure(figsize = (10,8))
plt.hist2d(kc.long, kc.lat, bins=100, cmap='hot')
plt.colorbar().set_label('Number of properties')
plt.xlabel('Longitude', fontsize=14)
plt.ylabel('Latitude', fontsize=14)
plt.title('Number of properties in each area of King County, Washington', fontsize=17)
plt.show()

Price per square foot will be calculated by dividing price by sqft_living (i.e. the square footage of the house):

In [ ]:
kc['price_per_sqft'] = kc.price/kc.sqft_living

The descriptive statistics and histogram below show that the median ppsf is $\$$245. However, the distribution has a long right tail, with some much more expensive properties having a ppsf of up to $\$$810.

In [ ]:
kc.price_per_sqft.describe()

In [ ]:
kc.price_per_sqft.hist(bins=range(0, 851, 50))
plt.xlabel('Price per square foot ($)', fontsize=12)
plt.ylabel('Frequency', fontsize=12)
plt.title('Histogram of house prices per square foot in King County, Washington', fontsize=15)
plt.xlim(0,850)
plt.show()

The figure below uses longitude and latitude as the x and y axes, respectively, to plot a scatter graph of ppsf to show their geographical distribution. It demonstrates that the highest ppsf are found in Bellevue, and the lowest are generally found outside of Seattle to the south.

In [ ]:
plt.figure(figsize = (10,8))
plt.scatter(kc.long, kc.lat ,c=kc.price_per_sqft, cmap = 'hot', s=1, alpha=0.5)
plt.colorbar().set_label('Price per square foot ($)')
plt.xlabel('Longitude', fontsize=14)
plt.ylabel('Latitude', fontsize=14)
plt.title('House prices in King County, Washington', fontsize=17)
plt.show()

In order to calculate the proportional difference in ppsf between the cheapest and the most expensive area, zipcodes will be used to group properties by location. The graph below plots mean ppsf by zipcode.

In [ ]:
# Mean price per square foot for each zipcode
zip_ppsf = kc.groupby(['zipcode'])['price_per_sqft'].mean().sort_values(ascending=False)
zip_ppsf.plot(kind='bar', figsize=(17,5))
plt.xlabel('Zipcode', fontsize=14)
plt.ylabel('Mean price per square foot ($)', fontsize=14)
plt.title('Mean prices per square foot in each zipcode in King County, Washington', fontsize=17)
plt.show()

The graph above shows that the most expensive zipcode is 98039 (Medina, home of tech billionaires), and the cheapest is 98023 (Federal Way). The proportional difference between the average ppsf in these two zipcodes is calculated below. **The price per square foot in Medina is 3.8 times that in Federal Way.**

In [ ]:
1+(zip_ppsf.max() - zip_ppsf.min())/zip_ppsf.min() # Calculating the increase from the cheapest to the most expensive

### Question 2: how much does having a view increase house price by?/which has a bigger impact on house price - being on the waterfront, or having a good view?

# COME BACK AND EDIT - CHOOSE QUESTION

The table and histogram below show the mean ppsf for different view ratings.

In [ ]:
kc.view.value_counts()

In [ ]:
kc.groupby(['view'])['price_per_sqft'].mean()

In [ ]:
# Mean price per square foot for each view rating
kc.groupby(['view'])['price_per_sqft'].mean().plot(kind='bar', figsize=(6,5), color=['grey', 'goldenrod', 'gold', 'yellowgreen', 'darkgreen'])
plt.xlabel('View rating', fontsize=14)
plt.xticks(rotation='horizontal')
plt.ylabel('Mean price per square foot ($)', fontsize=14)
plt.title('Mean prices per square foot for properties of different view ratings in King County, Washington', fontsize=17)
plt.show()

The analysis above shows that there is a considerable increase (25%) in mean ppsf from properties with 0 view (mean ppsf $\$$257) to a view rating of 1 (mean ppsf $\$$320). There is also a considerable increase (35%) from properties with a view rating of 3 ($\$$323) to a view rating of 4 ($\$$435). However, properties with view ratings of 1-3 achieve similar ppsft (between $\$$304 and $\$$323). There is even a slight dip from a view rating of 1 to 2. This could be caused by multiple other factors which influence house price. For example, it might be that a lot of properties with a rating of 2 are in a more urban area with smaller house sizes. Overall, the results indicate that there is a big difference between having no view and having some view, and between having a good view and having the best view, but that otherwise there is not much payoff in terms of increasing house prices by having different values or types of good view.

# BELOW SECTION IS WATERFRONT - EITHER EDIT OR REMOVE

In [ ]:
kc.groupby(['waterfront'])['price_per_sqft'].mean()

In [ ]:
# Mean price per square foot for each zipcode
kc.groupby(['waterfront'])['price_per_sqft'].mean().plot(kind='bar', figsize=(6,5), color=['mediumseagreen', 'royalblue'])
plt.xlabel('')
plt.xticks(np.arange(2), ('not on waterfront', 'waterfront'), rotation='horizontal', fontsize=14)
plt.ylabel('Mean price per square foot ($)', fontsize=14)
plt.title('Mean prices per square foot for properties on and away from the waterfront in King County, Washington', fontsize=17)
plt.show()

In [ ]:
kc.groupby(['waterfront', 'view'])['price_per_sqft'].mean()

In [ ]:
kc.groupby(['waterfront', 'view'])['price_per_sqft'].count()

In [ ]:
kc.groupby(['waterfront', 'view'])['price_per_sqft'].mean().plot(kind='bar')

***
# Modeling the data

This section will use simple and multiple linear regression to attempt to use the dataset to predice house prices. First, linear regression is conducted for each variable. Then multiple regression using a feature selection algorithm will be used. Test-train splits will be used to assess the prediction ability of the resulting model/s.

In the 'Exploring the data' section above, when testing the normality assumption, a table of model parameters was produced for the results of simple linear regression analysis run between each variable and the target variable of price:

In [ ]:
results

From the results of this analysis, the top 5 variables with the highest predictive power (i.e. highest r-squared values) when considered individually are grade, sqft_living, sqft_living15, sqft_above and bathrooms (below). I.e. the construction quality, house size total/above ground, house size of neighbouring houses, and number of bathrooms are the best individual predictors of house price. However, when taken alone, no single variable can explain more than 50% of the variation in price.

In [ ]:
results.sort_values(by=['r_squared'], ascending=False).head() # Top five highest r squared values for predictor variables

# COME BACK TO THE TWO CODE SECTIONS BELOW TO FIX MATRIC SCATTERPLOT/MAKE MORE EFFICIENT

In [ ]:
#fig, ax = plt.subplots(2,2)
#ax1, ax2, ax3, ax4 = ax.flatten()
#for n, col in enumerate(['sqft_living', 'sqft_living15', 'grade', 'bathrooms']):
#    axis = "ax" + str(n+1)
#    sns.regplot(kc_log[col], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=axis)
#    plt.legend()
#    plt.show()

In [ ]:
fig, ax = plt.subplots(3,5, figsize=(15,12))
ax1, ax2, ax3, ax4, ax5, ax6, ax7, ax8, ax9, ax10, ax11, ax12, ax13, ax14, ax15 = ax.flatten()
sns.regplot(kc_log['bedrooms'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax1)
sns.regplot(kc_log['bathrooms'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax2)
sns.regplot(kc_log['sqft_living'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax3)
sns.regplot(kc_log['sqft_lot'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax4)
sns.regplot(kc_log['floors'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax5)
sns.regplot(kc_log['view'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax6)
sns.regplot(kc_log['condition'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax7)
sns.regplot(kc_log['grade'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax8)
sns.regplot(kc_log['sqft_above'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax9)
sns.regplot(kc_log['sqft_basement'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax10)
sns.regplot(kc_log['yr_built'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax11)
sns.regplot(kc_log['lat'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax12)
sns.regplot(kc_log['long'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax13)
sns.regplot(kc_log['sqft_living15'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax14)
sns.regplot(kc_log['sqft_lot15'], kc_log['price'], label=col, marker = '.', scatter_kws={'s':1}, ax=ax15)
plt.show()

For modelling:
- Possible columns to create: garden size (lot size - living space size)

Iterating the model:
- Yr_built may have been more normal before it was log-transformed